File: swift-interpreter/Sources/Interpreter/Lexer.swift
public struct Lexer {
	let input: String
	var position: Int
	var ch: Character?

	public init(_ input: String) {
		self.input = input
		self.position = 0
		self.ch = !input.isEmpty ? input.first : nil  
	}

	public mutating func nextToken() -> Token {	
		self.skipWhitespace()
		let token: Token
		switch self.ch {
			case nil : token = .eof
			case "=" : token = self.hasNext(ch: "=", isToken: .eq, elseIsToken: .assign)
			case "+" : token = .plus
			case "(" : token = .lparen
			case ")" : token = .rparen
			case "{" : token = .lcurly
			case "}" : token = .rcurly
			case "," : token = .comma
			case ";" : token = .semicolon
			case "-" : token = .minus
			case "!" : token = self.hasNext(ch: "=", isToken: .notEq, elseIsToken: .bang)
			case "*" : token = .asterisk
			case "/" : token = .slash
			case "<" : token = .lt
			case ">" : token = .gt
			case let .some(ch) where ch.isLetter : return readIdent()
			case let .some(ch) where ch.isNumber : return readDigit()
			default: token = .illegal
		}
		
		self.advance()
		return token
	}

	mutating func advance() {
		let nextCh = peekNext()
		if nextCh != nil {
			self.position += 1
		}
		self.ch = nextCh
	}

	func peekNext() -> Character? {
		let readPosition = self.position + 1
  	if readPosition >= self.input.count {
			return nil
		}
		let index = self.input.index(self.input.startIndex, offsetBy: readPosition)
		return self.input[index];
	}

	mutating func hasNext(ch: Character, isToken: Token, elseIsToken: Token) -> Token {
		if self.peekNext() == ch {
			self.advance()
			return isToken
		} else {
			return elseIsToken;
		}
	}

	mutating func readWhile(_ pred: (Character) -> Bool) -> String {
		var str = String.init()
		while let ch = self.ch, pred(ch) {
			str += String(ch)
			self.advance()
		}
		return str
	}

	mutating func skipWhitespace() {
		let _ = readWhile({ $0.isWhitespace })
	}

	mutating func readIdent() -> Token {
		let ident = readWhile({ $0.isLetter })
		return lookupIdent(ident)
	}

	mutating func readDigit() -> Token {
		let digit = readWhile({ $0.isNumber })
		return Int(digit).map(Token.int) ?? .illegal
	}
}

File: ocaml-interpreter/lib/lexer.ml
open Base

type lexer = { input : string; position : int; ch : char option }

let make input =
  if String.is_empty input then
    { input; position = 0; ch = None }
  else
    { input; position = 0; ch = Some (String.get input 0) }
  
let rec next_token lexer =
  let open Token in
  let lexer = skip_whitespace lexer in
  match lexer.ch with
    | None -> Eof, lexer
    | Some ch -> match ch with
      | ch when is_ident ch -> read_ident lexer
      | ch when is_number ch -> read_number lexer
      | _ -> read_symbol lexer

and advance lexer =
  let ch = peek_next lexer in
  match ch with
  | None -> { lexer with ch }
  | ch -> { lexer with position= (inc lexer.position); ch }

and peek_next lexer =
  if inc lexer.position >= String.length lexer.input
  then None
  else Some (String.get lexer.input (inc lexer.position))

and match_next lexer value matched default =
  let token, lexer = match peek_next lexer with
  | Some next when Char.(next = value) -> matched, advance lexer
  | _ -> default, lexer
  in
  token, advance lexer
  
and read_ident lexer =
  let lexer, ident = advance_while lexer is_ident in
  Token.lookup_ident ident, lexer

and read_number lexer = 
  let lexer, number = advance_while lexer is_number in
  Token.Int (Int.of_string number), lexer 

and read_symbol lexer = 
  let open Token in
  match lexer.ch with
    | None -> Eof, lexer
    | Some ch -> 
      let token, lexer = match ch with 
      | '=' -> match_next lexer '=' Equal Assign
      | '+' -> Plus, advance lexer
      | '(' -> LeftParen, advance lexer
      | ')' -> RightParen, advance lexer
      | '{' -> LeftCurly, advance lexer
      | '}' -> RightCurly, advance lexer
      | ',' -> Comma, advance lexer
      | ';' -> Semicolon, advance lexer
      | '-' -> Minus, advance lexer
      | '!' -> match_next lexer '=' NotEqual Bang
      | '*' -> Asterisk, advance lexer
      | '/' -> Slash, advance lexer
      | '<' -> LessThan, advance lexer
      | '>' -> GreaterThan, advance lexer
      | _ -> Illegal, advance lexer
    in
    token, lexer
  
and skip_whitespace lexer =
  let lexer, _ = advance_while lexer is_whitespace in
  lexer

and advance_while lexer pred =
  let rec advance_while_aux acc lexer =
    match lexer.ch with
    | None -> lexer, acc
    | Some ch -> if pred ch
      then advance_while_aux (acc ^ String.make 1 ch) (advance lexer)
      else lexer, acc
  in
  advance_while_aux Caml.String.empty lexer

and is_ident ch = 
  Char.(is_alpha ch || ch = '_')
and is_number = Char.is_digit
and is_whitespace = Char.is_whitespace
and inc = Int.succ